hydro_deploy/core/src/ssh.rs

```diff
@@ -88,65 +92,103 @@ impl LaunchedBinary for LaunchedSshBinary {
+                // `exit_status()`
+                self.channel.wait_close().await?;
             }
 
-            // TODO(mingwei): re-use this code to download `dtrace` data.
-            // let sftp = self.session.as_ref().unwrap().sftp().await?;
-            // // Check if perf.data exists
-            // match sftp.open(&PathBuf::from(output_file)).await {
-            //     Ok(perf_data) => {
-            //         // download perf.data
-            //         let mut perf_data_local = tokio::fs::File::create(output_file).await?;
-            //         tokio::io::copy(&mut perf_data, &mut perf_data_local).await?;
-            //     }
-            //     Err(Error::Ssh2(ssh2_err)) if ErrorCode::SFTP(2) == ssh2_err.code() => {
-            //         // File not found, probably due to some other previous error.
-            //         ProgressTracker::println(&format!(
-            //             "perf output file {:?} not found on remote machine (SFTP error 2).",
-            //             output_file
-            //         ));
-            //     }
-            //     Err(unexpected_err) => Err(unexpected_err)?,
-            // }
-        };
+            // Run perf post-processing and download perf output.
+            if let Some(perf) = &mut self.perf {
+                {
+                    let set_msg = &set_msg;
+                    let set_progress = &set_progress;
+                    (set_msg)("perf script".to_owned());
+                    (set_progress)(10);
+
+                    let mut script_channel =
+                        self.session.as_ref().unwrap().channel_session().await?;
+                    let mut stderr_lines = FuturesBufReader::new(script_channel.stderr()).lines();
+                    let stdout = script_channel.stream(0);
+                    let mut fold_er = Folder::from(perf.fold_options.clone().unwrap_or_default());
+                    let (fold_send, graph_recv) = tokio::io::duplex(8 * 1024);
+                    let local_file =
+                        tokio::fs::File::create(perf.output_file.to_str().unwrap()).await?;
+
+                    // Pattern on `()` to make sure no `Result`s are ignored.
+                    let ((), (), (), ()) = tokio::try_join!(
+                        async move {
+                            // Log stderr.
+                            while let Some(Ok(s)) = stderr_lines.next().await {
+                                ProgressTracker::println(&format!("[perf stderr] {s}"));
+                            }
+                            Ok(()) as Result<_>
+                        },
+                        async move {
+                            // Download perf output and fold.
+                            tokio::task::spawn_blocking(move || {
+                                fold_er.collapse(
+                                    SyncIoBridge::new(TokioBufReader::new(stdout)),
+                                    SyncIoBridge::new(fold_send),
```

Comment:
I think it may be nice to store the output of fold in memory, and then have a separate API to generate the flamegraph from that. Reason I'm thinking of this is because the folded output is useful on its own to be loaded into something like speedscope.app, which has a pretty nice interactive UI.

Reply:
Added a `fold_outfile` optional config